--- u-boot-2011.12-2013_Q2.1/arch/arm/cpu/armv7/start.S	2013-09-03 01:19:34.000000000 +0300
+++ u-boot-2011.12-2013_Q2.1-be/arch/arm/cpu/armv7/start.S	2014-01-26 20:29:57.701224074 +0200
@@ -126,6 +126,10 @@
 #ifdef CONFIG_MARVELL
 @ Initial PJ4B configurations
 @ configure Marvell debug R1
+#if defined(__BE) && defined(CPU_ARMARCH7)
+	SETEND BE /* Sets ENDIANSTATE to 1, for big-endian operation */
+#endif
+
 	mrc 	p15, 1, r1, c15, c1, 1
 	orr	r1, r1, #0x00010000	@ BIT16: Disable Data transfer for clean line
 	orr	r1, r1, #0x00000080	@ BIT7: Enable KW IO Implementation Specific
@@ -150,6 +154,9 @@
 	/* Set Registers Base address */
 	ldr  r1, =INTER_REGS_BASE
 	ldr  r0, =0xD0020080
+#if defined(__BE) && defined(CPU_ARMARCH7)
+	rev r1, r1
+#endif
 	str  r1, [r0]
 
 	/* notify kernel on 0xf1 switch */
@@ -167,10 +174,7 @@
 @ Initial PJ4B configurations
 @ configure Marvell debug R1
 */
-#if defined(__BE)
-#if defined(CPU_ARMARCH7)
-		SETEND BE /* Sets ENDIANSTATE to 1, for big-endian operation */
-#else
+#if defined(__BE) && !defined(CPU_ARMARCH7)
 	/* disable I-Cache */
 	.word 0x100f11ee /* mrc 	p15, 0, r0, c1, c0 */
 	.word 0x010ac0e3 /* bic     r0, r0, #4096   ; 0x1000 I-Cache Enable bit*/
@@ -178,7 +182,6 @@
 	.word 0x020080e3 /* orr     r0, r0, #2      ; 0x2 */
 	.word 0x800080e3 /* orr 	r0, r0, #0x80 	; Endianness = big endians*/
 	.word 0x100f01ee /* mcr     15, 0, r0, cr1, cr0, {0} */
-#endif
 	nop;nop;nop;nop;
 	nop;nop;nop;nop;
 #endif
